### PR TITLE
add contexts to scope and apply it

### PR DESCRIPTION
**Pull request by rxlabz**
## :loudspeaker: Type of change

- [ ] Bugfix
- [ ] New feature
- [x] Enhancement
- [ ] Refactoring

## :scroll: Description

add a `contexts` property to `Scope`

## :bulb: Motivation and Context

https://develop.sentry.dev/sdk/unified-api/

## :pencil: Checklist

- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [x] All tests passing
- [x] No breaking changes

### CODE REVIEW

File: dart/lib/src/protocol/contexts.dart [hunk 1]
```diff
@@ -11,107 +13,170 @@ import 'runtime.dart';
...
-class Contexts {
-  const Contexts({
-    this.device,
-    this.operatingSystem,
-    this.runtimes,
-    this.app,
-    this.browser,
-    this.gpu,
-  });
+class Contexts extends MapView {
```
Comment (marandaneto) [inline dart/lib/src/protocol/contexts.dart]: just out of curiosity, why not `HashMap`? what's the difference between them?

File: dart/lib/src/protocol/contexts.dart [hunk 2]
```diff
@@ -11,107 +13,170 @@ import 'runtime.dart';
...
-              while (json.containsKey(key)) {
-                key = '$key$k';
-                k++;
+    forEach((key, value) {
+      if (value == null) return;
+      switch (key) {
+        case Device.type:
+          Map<String, dynamic> deviceMap;
+          if ((deviceMap = device.toJson()).isNotEmpty) {
+            json['device'] = deviceMap;
```
Comment (marandaneto) [inline dart/lib/src/protocol/contexts.dart]:
```suggestion
            json[Device.type] = deviceMap;
```

wouldd this work too?

File: dart/lib/src/protocol/contexts.dart [hunk 3]
```diff
@@ -11,107 +13,170 @@ import 'runtime.dart';
...
+          break;
+
+        case App.type:
+          Map<String, dynamic> appMap;
+          if ((appMap = app.toJson()).isNotEmpty) {
+            json[App.type] = appMap;
+          }
+          break;
+
+        case 'browser':
```
Comment (marandaneto) [inline dart/lib/src/protocol/contexts.dart]: why do we use the `Class.type ` on the case above and here a plain String?

File: dart/lib/src/protocol/contexts.dart [hunk 4]
```diff
@@ -11,107 +13,170 @@ import 'runtime.dart';
...
+            json['gpu'] = gpuMap;
+          }
+          break;
+
+        case 'runtimes':
+          if (runtimes != null) {
+            if (runtimes.length == 1) {
+              final runtime = runtimes[0];
+              if (runtime != null) {
+                final key = runtime.key ?? 'runtime';
```
Comment (marandaneto) [inline dart/lib/src/protocol/contexts.dart]: Runtime.type?

File: dart/lib/src/protocol/contexts.dart [hunk 5]
```diff
@@ -11,107 +13,170 @@ import 'runtime.dart';
...
+          json[key] = value;
       }
-    }
+    });
 
     return json;
   }
+
+  Contexts clone() {
+    // TODO ? make each class cloneable ?
```
Comment (marandaneto) [inline dart/lib/src/protocol/contexts.dart]: is this related to this? https://github.com/getsentry/sentry-dart/projects/1#card-47865684
if so, we can remove the comment, we already have a card for it

File: dart/test/scope_test.dart [hunk 6]
```diff
@@ -2,6 +2,8 @@ import 'package:collection/collection.dart';
 import 'package:sentry/sentry.dart';
 import 'package:test/test.dart';
 
+// TODO test context apply / clone
```
Comment (marandaneto) [inline dart/test/scope_test.dart]: is this part of this PR?

**Comment by marandaneto** [comment]
this PR is not addressing the SentryClient changes:
which is merging the Event Contexts and the Scope Contexts
https://github.com/getsentry/sentry-java/blob/5ebd49cbfd17eb66175e9247db6d27e4df84a56b/sentry/src/main/java/io/sentry/SentryClient.java#L317-L321

**Comment by rxlabz** [comment]
> this PR is not addressing the SentryClient changes:
> which is merging the Event Contexts and the Scope Contexts

@marandaneto  is this incorrect or incomplete? https://github.com/getsentry/sentry-dart/blob/262305a074a3d0ce497cc0811b221f252a9fcdef/dart/lib/src/scope.dart#L162-L166

File: dart/test/scope_test.dart (hunk 6 above)
Comment (rxlabz) [reply dart/test/scope_test.dart]: ✅

File: dart/lib/src/protocol/contexts.dart (hunk 5 above)
Comment (rxlabz) [reply dart/lib/src/protocol/contexts.dart]: ✅

File: dart/lib/src/protocol/contexts.dart (hunk 4 above)
Comment (rxlabz) [reply dart/lib/src/protocol/contexts.dart]: ✅

File: dart/lib/src/protocol/contexts.dart (hunk 3 above)
Comment (rxlabz) [reply dart/lib/src/protocol/contexts.dart]: ✅

File: dart/lib/src/protocol/contexts.dart (hunk 2 above)
Comment (rxlabz) [reply dart/lib/src/protocol/contexts.dart]: ✅

File: dart/lib/src/protocol/contexts.dart (hunk 1 above)
Comment (rxlabz) [reply dart/lib/src/protocol/contexts.dart]: Extending HashMap requires implementing most of its members. MapView is only a wrapper around an initial Map, which seemed simpler in this case.

Comment (marandaneto) [reply dart/lib/src/protocol/contexts.dart]: I see, thanks.
Should we at least set its type then? `extends MapView<String, dynamic> `

File: dart/lib/src/scope.dart [hunk 7]
```diff
@@ -53,7 +53,13 @@ class Scope {
 
   Map<String, dynamic> get extra => Map.unmodifiable(_extra);
 
-  // TODO: Contexts
+  final Contexts _contexts = Contexts();
+
+  Contexts get contexts => _contexts;
+
+  void setContexts({String key, dynamic value}) {
```
Comment (marandaneto) [inline dart/lib/src/scope.dart]: does it need to be named? because the other add methods are not, only this one will be different

Comment (marandaneto) [reply dart/lib/src/scope.dart]: missing method `removeContexts`

Comment (marandaneto) [reply dart/lib/src/scope.dart]: also, `setContexts` can be `dynamic` as value, but our protocol does not accept non objects here, see what is done on Java, we'd need something similar or `setContexts('a', 'b')` would cause a backend issue

https://github.com/getsentry/sentry-java/blob/main/sentry/src/main/java/io/sentry/Scope.java#L340-L374

File: dart/test/scope_test.dart [hunk 8]
```diff
@@ -222,6 +222,42 @@ void main() {
...
+      ..fingerprint = ['example-dart']
+      ..addBreadcrumb(breadcrumb)
+      ..transaction = '/example/app'
+      ..level = SentryLevel.warning
+      ..setTag('build', '579')
+      ..setExtra('company-name', 'Dart Inc')
+      ..setContexts(key: 'theme', value: 'material')
+      ..addEventProcessor(
+        (event, hint) => event..tags.addAll({'page-locale': 'en-us'}),
+      );
```
Comment (marandaneto) [inline dart/test/scope_test.dart]: please add a few values to the key/value objects to test the precedence over others, eg
a context key on the event cant be replaced by a value of the same key on the scope.

for the level, this is actually the opposite, the scope level has precedence over the event level

**Comment by marandaneto** [comment]
> > this PR is not addressing the SentryClient changes:
> > which is merging the Event Contexts and the Scope Contexts
> 
> @marandaneto is this incorrect or incomplete?
> 
> https://github.com/getsentry/sentry-dart/blob/262305a074a3d0ce497cc0811b221f252a9fcdef/dart/lib/src/scope.dart#L162-L166

more to incomplete than incorrect, I've made a few new comments

File: dart/lib/src/protocol/runtime.dart [hunk 9]
```diff
@@ -1,9 +1,13 @@
+const runtimesType = 'runtimes';
```
Comment (marandaneto) [inline dart/lib/src/protocol/runtime.dart]: just out of curiosity, why the type is not in the class?

File: dart/lib/src/scope.dart [hunk 10]
```diff
@@ -53,7 +53,13 @@ class Scope {
 
   Map<String, dynamic> get extra => Map.unmodifiable(_extra);
 
-  // TODO: Contexts
+  final Contexts _contexts = Contexts();
+
+  Contexts get contexts => _contexts;
```
Comment (marandaneto) [inline dart/lib/src/scope.dart]:
```suggestion
  Contexts get contexts => Map.unmodifiable(_contexts);
```

File: dart/lib/src/protocol/contexts.dart (hunk 1 above)
Comment (rxlabz) [reply dart/lib/src/protocol/contexts.dart]: ✅

File: dart/lib/src/scope.dart (hunk 7 above)
Comment (rxlabz) [reply dart/lib/src/scope.dart]: - `setContexts` args are now mandatory 
- added removeContexts
- I added some conditions to the setter ( not null and of type `num`, `bool` or `String` )
 What would be the best strategy in case of invalid value? For now, it's a no_op.

Comment (rxlabz) [reply dart/lib/src/scope.dart]: Also, now, the value must be an Object ( with a `toString()` method ) so we could just set the toString() if the value is not a num, a bool, or a String?
```dart
if (key != null && value != null) return;

    _contexts[key] = (value is num || value is bool || value is String) ? value : '$value';
```

File: dart/lib/src/scope.dart (hunk 10 above)
Comment (rxlabz) [reply dart/lib/src/scope.dart]: ✅

File: dart/lib/src/protocol/runtime.dart (hunk 9 above)
Comment (rxlabz) [reply dart/lib/src/protocol/runtime.dart]: ✅ I moved it in the class and renamed it `Runtime.listType`

File: dart/lib/src/scope.dart (hunk 7 above)
Comment (marandaneto) [reply dart/lib/src/scope.dart]: @rxlabz on Java, we'd not return, but rather wrap into a `value` field if they are primitive types.
This deviates quite a lot of the Java impl. and limit the API quite a lot, I'd suggest doing something similar to Java.

exposing more setContexts methods or increment the current solution.

_contexts[key] = (value is num || value is bool || value is String) ? [{'value': value] : value;

sounds more realistic

File: dart/test/scope_test.dart (hunk 8 above)
Comment (rxlabz) [reply dart/test/scope_test.dart]: ✅

File: dart/lib/src/scope.dart (hunk 7 above)
Comment (rxlabz) [reply dart/lib/src/scope.dart]: sorry, I missed the java `public void setContexts(final @NotNull String key, final @NotNull Object value)` and misunderstood the Api.

File: dart/test/scope_test.dart [hunk 11]
```diff
@@ -230,12 +230,90 @@ void main() {
...
+      id: '800',
+      username: 'first-user',
+      email: 'first@user.lan',
+      ipAddress: '127.0.0.1',
+      extras: <String, String>{'first-sign-in': '2020-01-01'},
+    );
+
+    final breadcrumb = Breadcrumb(message: 'Authenticated');
+
+    setUp(() {});
```
Comment (marandaneto) [inline dart/test/scope_test.dart]: do we need this?

File: dart/test/scope_test.dart [hunk 12]
```diff
@@ -230,12 +230,90 @@ void main() {
...
+      expect(updatedEvent.breadcrumbs, [breadcrumb]);
+      expect(updatedEvent.level, SentryLevel.warning);
+      expect(updatedEvent.tags,
+          {'etag': '987', 'build': '579', 'page-locale': 'en-us'});
+      expect(
+          updatedEvent.extra, {'e-infos': 'abc', 'company-name': 'Dart Inc'});
+      expect(updatedEvent.contexts['theme'], 'material');
+    });
+
+    test('should not apply context properties when event has already it ', () {
```
Comment (marandaneto) [inline dart/test/scope_test.dart]:
```suggestion
    test('should not apply context properties when event already has it ', () {
```

Comment (marandaneto) [reply dart/test/scope_test.dart]: I'd make this test smaller and more focused on the context field, right now all the user, breadcrumb etc is just noisy for the name of the test, the test `apply context to event` already cover all the other fields, right

Comment (marandaneto) [reply dart/test/scope_test.dart]: I dont see this test testing the `contexts` field though, its not even called

File: dart/lib/src/protocol/contexts.dart [hunk 13]
```diff
@@ -11,107 +13,169 @@ import 'runtime.dart';
...
+    final copy = Contexts(
+      device: device?.clone(),
+      operatingSystem: operatingSystem?.clone(),
+      app: app?.clone(),
+      browser: browser?.clone(),
+      gpu: gpu?.clone(),
+      runtimes: runtimes.map((runtime) => runtime.clone()).toList(),
+    )..addEntries(
+        entries.where((element) => !_defaultFields.contains(element.key)),
+      );
```
Comment (marandaneto) [inline dart/lib/src/protocol/contexts.dart:167]: is this case covered by an unit test?

Comment (rxlabz) [reply dart/lib/src/protocol/contexts.dart:167]: is it enough? https://github.com/getsentry/sentry-dart/blob/99027322f429d6884cf7ec9a85a2053426ffef49/dart/test/contexts_test.dart#L127-L128

File: dart/test/scope_test.dart (hunk 12 above)
Comment (rxlabz) [reply dart/test/scope_test.dart]: ✅

File: dart/test/scope_test.dart (hunk 11 above)
Comment (rxlabz) [reply dart/test/scope_test.dart]: ✅

File: dart/lib/src/scope.dart (hunk 7 above)
Comment (rxlabz) [reply dart/lib/src/scope.dart]: ✅

File: dart/lib/src/scope.dart [hunk 14]
```diff
@@ -153,6 +171,17 @@ class Scope {
...
 
+    _contexts.clone().forEach((key, value) {
+      // add the runtime list to the event.contexts.runtimes
+      if (key == Runtime.listType && value is List && value.isNotEmpty) {
+        value.forEach((runtime) => event.contexts.addRuntime(runtime));
+      } else if ((!event.contexts.containsKey(key) ||
+              event.contexts[key] == null) &&
+          value != null) {
+        event.contexts[key] = value;
+      }
```
Comment (marandaneto) [inline dart/lib/src/scope.dart]: why do we need to special case here? its not just adding all the keys from a map to another map?

Comment (rxlabz) [reply dart/lib/src/scope.dart]: should we override the event.contexts.runtimes ?

Comment (rxlabz) [reply dart/lib/src/scope.dart]: In Java, it's not a list so it makes sense to override but with a list, I wasn't sure.

Comment (marandaneto) [reply dart/lib/src/scope.dart]: we should probably merge them, yeah.
can you do a method like you do with `_mergeEventExtra` or `_mergeEventExtra`? seems easier to read (to merge runtimes).

also, we dont allow null keys right, so `event.contexts[key] == null` seems unecessary?

Comment (rxlabz) [reply dart/lib/src/scope.dart]: ✅

Comment (rxlabz) [reply dart/lib/src/scope.dart]: for now the `Contexts` ctor allows null values 
```dart
Contexts({
    Device device,
    OperatingSystem operatingSystem,
    List<Runtime> runtimes,
    App app,
    Browser browser,
    Gpu gpu,
  }) : super({
          Device.type: device,
          OperatingSystem.type: operatingSystem,
          Runtime.listType: runtimes ?? [],
          App.type: app,
          Browser.type: browser,
          Gpu.type: gpu,
        });
```

Comment (marandaneto) [reply dart/lib/src/scope.dart]: true, good catch

**Review by marandaneto — APPROVED** [review]
LGTM
